ledger/store/trackerdb/dualdriver/dualdriver.go

```diff
@@ -0,0 +1,518 @@
+// Close implements trackerdb.Transaction
+func (tx *transaction) Close() error {
+	errP := tx.primary.Close()
+	errS := tx.secondary.Close()
+	// errors are unlikely to match between engines
+	return coalesceErrors(errP, errS)
+}
+
+type snapshot struct {
+	primary   trackerdb.Snapshot
+	secondary trackerdb.Snapshot
+	trackerdb.Reader
+}
+
+// ResetTransactionWarnDeadline implements trackerdb.Snapshot
+func (s *snapshot) ResetTransactionWarnDeadline(ctx context.Context, deadline time.Time) (prevDeadline time.Time, err error) {
+	_, _ = s.primary.ResetTransactionWarnDeadline(ctx, deadline)
+	_, _ = s.secondary.ResetTransactionWarnDeadline(ctx, deadline)
+	// ignore results, this is very engine specific
+	return
+}
+
+// Close implements trackerdb.Snapshot
+func (s *snapshot) Close() error {
+	errP := s.primary.Close()
+	errS := s.secondary.Close()
+	// errors are unlikely to match between engines
+	return coalesceErrors(errP, errS)
+}
+
+//
+// refs
+//
+
+type accountRef struct {
+	primary   trackerdb.AccountRef
+	secondary trackerdb.AccountRef
+}
+
+// AccountRefMarker implements trackerdb.AccountRef
+func (accountRef) AccountRefMarker() {}
+
+func coalesceAccountRefs(primary, secondary trackerdb.AccountRef) (trackerdb.AccountRef, error) {
+	if primary != nil && secondary != nil {
+		return accountRef{primary, secondary}, nil
+	} else if primary == nil && secondary == nil {
+		// all good, ref is nil
+		return nil, nil
+	} else {
+		// ref mismatch
+		return nil, ErrInconsistentResult
+	}
+}
+
+type onlineAccountRef struct {
+	primary   trackerdb.OnlineAccountRef
+	secondary trackerdb.OnlineAccountRef
+}
+
+// OnlineAccountRefMarker implements trackerdb.OnlineAccountRef
+func (onlineAccountRef) OnlineAccountRefMarker() {}
+
+func coalesceOnlineAccountRefs(primary, secondary trackerdb.OnlineAccountRef) (trackerdb.OnlineAccountRef, error) {
+	if primary != nil && secondary != nil {
+		return onlineAccountRef{primary, secondary}, nil
+	} else if primary == nil && secondary == nil {
+		// all good, ref is nil
+		return nil, nil
+	} else {
+		// ref mismatch
+		return nil, ErrInconsistentResult
+	}
+}
+
+type resourceRef struct {
+	primary   trackerdb.ResourceRef
+	secondary trackerdb.ResourceRef
+}
+
+// ResourceRefMarker implements trackerdb.ResourceRef
+func (resourceRef) ResourceRefMarker() {}
+
+type creatableRef struct {
+	primary   trackerdb.CreatableRef
+	secondary trackerdb.CreatableRef
+}
+
+// CreatableRefMarker implements trackerdb.CreatableRef
+func (creatableRef) CreatableRefMarker() {}
+
+//
+// helpers
+//
+
+func coalesceErrors(errP error, errS error) error {
+	if errP == nil && errS != nil {
+		logging.Base().Error("secondary engine error, ", errS)
+		return ErrInconsistentResult
+	}
+	if errP != nil && errS == nil {
+		logging.Base().Error("primary engine error, ", errP)
+		return ErrInconsistentResult
+	}
+	// happy case (no errors)
+	if errP == nil && errS == nil {
+		return nil
+	}
+	// happy case (both errored)
+	return errP
```

Comment:
It's possible for the two drivers to return fundamentally different errors, right? 

In that situation, it's valuable to pass back the errors of primary and secondary drivers, and maybe even inspect them before declaring they match.

The more I think about it, I worry this function could hide situations where the two errors are very different:
For example, you could have a test `TestErrorWriting` which on the Primary Driver fails for the reason you expect, and your Secondary Driver could fail for some unrelated reason, like your disk is full. In this case, `coalesceErrors` will see that both drivers had *some* error, and returns the `errP`, which is the error we're testing for. Using that result to declare the two databases consistent would be a mistake because we don't actually know what happened to the secondary driver. In other words, the test is quietly not testing what we expect.

I guess it just comes down to the breadth of reasons a given command could fail. If every place we call `coalesceErrors` is so tightly scoped that *any* error is functionally equivalent, then I guess this matters less.

Reply:
without creating a common error enum to be used by all engines for returning errors, the errors returned are likely to always be different. 

The only common error type that gets normalized on the implementation of the `sqlitedriver` and `pebbledbdriver` is `ErrNotFound`, everything else is an error bubbling up from sqlite or from pebbledb and this will never match. 

The alternative design here is to handle the `ErrNotFound` and forward it when found, then for all other errors have a `DualDriverError { primaryError, secondaryError }` just for debugging printouts on the test.

Reply:
Yeah, since the two engines are different, comparing their errors would be tricky. It's not even clear that the same ErrorX from both engines would mean the same thing.

Regardless, it'd be nice to be less optimistic here but I don't think it's required for this PR.